proposals/0000-hypercore-header.md

```diff
@@ -0,0 +1,109 @@
+
+
+# Motivation
+[motivation]: #motivation
+
+At time of writing, Dat's standard data structures are being expanded upon (hyperdb) while existing data structures are being updated with breaking changes (hyperdrive). In order for these changes to be deployed smoothly to the network, it's important that the following needs be met:
+
+ 1. Programs should correctly identify the data structure type and version within a hypercore. This will enable the program to use the correct handling code (eg hyperdb, hyperdrive-v1, hyperdrive-v2).
+ 2. Programs should abort processing a hypercore if they lack support for the type or version of the data structure.
+
+
+# Usage Documentation
+[usage-documentation]: #usage-documentation
+
+The "header" is the first entry in a hypercore. It includes a `type` and an optional `extension` which can contain data-structure-specific data.
+
+A program that is reading a hypercore will examine the `type` to determine how to process the hypercore.
+
+```js
+function loadCorrectStructure (hypercore, cb) {
+  readHeader(hypercore, (err, header) => {
+    if (err) return cb(err)
+
+    if (!header) {
+      // no header present - treat as a hypercore
+      return cb(null, hypercore)
+    }
+
+    switch (header.type) {
+      case 'hyperdrive':
+        return cb(null, createHyperdriveV1(hypercore, header))
+      case 'hyperdrive-v2':
+        return cb(null, createHyperdriveV2(hypercore, header))
+      case 'hyperdb':
+        return cb(null, createHyperdbV1(hypercore, header))
+      case '':
+        return cb(null, hypercore) // no structure, treat as a hypercore
+      // ...
+      default:
+        return cb(new Error('Unsupported: ' + header.type))
+    }
+  })
+}
+```
+
+The `type` string should be unique to the data structure (refer to existing DEPs to avoid conflicts). When a breaking change is made to the data-structure, it should be given a new `type` string. For example, the `"hyperdrive"` type string will be updated to `"hyperdrive-v2"` for its second version.
+
+The `header.extension` is an optional blob of bytes. It can be used by the data structure to store custom data. In `"hyperdrive"` (v1) it is used to specify the key of the content hypercore. It is also valid to encode a protobuf message into the extension field, and therefore it's simple to add additional fields.
+
+
+# Reference Documentation
+[reference-documentation]: #reference-documentation
+
+The header message should be written as the first entry of the hypercore. It is not required that all hypercores possess a header.
+
+The header message has the following protobuf schema:
+
+```protobuf
+message HypercoreHeader {
+  required string type = 1;
+  optional bytes extension = 2;
+}
+```
+
+Data structures should not add additional fields to the `HypercoreHeader`. They should put any additional data into the `extension`. This is easy to do with protobuf's nested messages. For example:
```

Comment:
Should implementations fail fast on additional fields, or should they ignore additional fields? The later is future-compatible, the former is not.

Reply:
Ignore, probably. I'll add a note.